### PR TITLE
ENG-51 Use nta.utils retry decorator in numeta-apps

### PR DESCRIPTION
**Pull request by SaganBolliger**
Fixes https://jira.numenta.com/browse/ENG-51

### CODE REVIEW

**Comment by oxtopus** [comment]
LGTM :+1:

**Comment by oxtopus** [comment]
This is causing an error in the result quality test:

```
________________________ ResultQualityTests.testRNSData ________________________
self = <tests.integration.result_quality_test.ResultQualityTests testMethod=testRNSData>

    def testRNSData(self):
      """
        Tests for expected result quality from RNS data.
    
        4030 rows
        """
    
      dataIdentifier = "RNS"
      knownDataFile = "rns_Backend_InstanceId=i-57009237_DiskWriteOps.csv"
      expectedResults = {"fn": 3,
                         "fp": 8,
                         "tn": 3683,
                         "tp": 2,
                         "quality": -230}
    
>     self._runQualityTest(dataIdentifier, knownDataFile, expectedResults)

../htmengine/tests/integration/result_quality_test.py:171: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.integration.result_quality_test.ResultQualityTests testMethod=testRNSData>
dataIdentifier = 'RNS'
knownDataFile = 'rns_Backend_InstanceId=i-57009237_DiskWriteOps.csv'
expectedResults = {'fn': 3, 'fp': 8, 'quality': -230, 'tn': 3683, ...}

    def _runQualityTest(self, dataIdentifier, knownDataFile, expectedResults):
      """
        Runs the data from knownDataFile (a csv) through htmengine and verifies
        htmengine returns the expected values in terms of a confusion matrix dict
        expectedResults.
    
        :param dataIdentifier: A string to identify this data in logs
        :param knownDataFile: A csv filename that exist in local data/ dir
        :param expectedResults: The confusion matrix and quality score we expect
                                out of htmengine.
        :type expectedResults: dict
    
        :returns: A sequence of results from the model; each result is a sequence of
          <Date string>, <Value>, <Likelihood Score>,  <Record Number>
        """
    
      metricName = self._genUniqueMetricName(dataIdentifier)
    
      self.addCleanup(self._deleteMetric, metricName)
    
      LOGGER.info(self.initialLoggingString, metricName)
    
      # Get path to data
      knownDataFilePath = self._getPathToData(knownDataFile)
    
      # Load and send data
      sock = self._getSocketConnection()
      LOGGER.info("Sending data from %s ...", knownDataFile)
      labels = self._loadAndSendData(sock, knownDataFilePath, metricName)
    
      # Make sure the metric was properly created and wait for the expected
      # records to be stored. NOTE: Waiting for all records to be stored
      # facilitates constistent stats calculation in htmengine, resulting in
      # consistency of results from one run of the test to the next.
      uid = self.checkMetricCreated(metricName, numRecords=len(labels))
    
      # Save the uid for later
      LOGGER.info("Metric %s has uid: %s", metricName, uid)
    
      # Send model creation request
      nativeMetric = {"datasource": "custom",
                      "metricSpec": {"uid": uid}}
    
      model = self._createModel(nativeMetric)
      self.assertEqual(model.uid, uid)
      self.assertEqual(model.name, metricName)
      self.assertEqual(model.server, metricName)
    
      return self._verifyResults(uid, metricName, knownDataFilePath, labels,
>                                expectedResults)

../htmengine/tests/integration/result_quality_test.py:525: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.integration.result_quality_test.ResultQualityTests testMethod=testRNSData>
uid = 'ed0da9015f504a9fb6c8169a2e99294e'
metricName = 'test.metric.RNS.1444941021.87'
knownDataFilePath = '/opt/numenta/products/htmengine/tests/integration/data/rns_Backend_InstanceId=i-57009237_DiskWriteOps.csv'
labels = [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, ...]
expectedResults = {'fn': 3, 'fp': 8, 'quality': -230, 'tn': 3683, ...}
boundingRange = 0.1

    def _verifyResults(self,
                       uid,
                       metricName,
                       knownDataFilePath,
                       labels,
                       expectedResults,
                       boundingRange = .1):
      """
        Waits for the model to complete running through the data up to lastRowId
        and then computes a confusion matrix over all the results and
        compares those values to the pre-calculated expectedResults.
    
        :param uid: The uid of the metric / model
        :param metricName: name of custom metric
        :param lastRowId: The count of rows in the data / expected count of results
        :param labels: The ground truth labels for each row of data.
        :param expectedResults: A dict containing pre-computed confusion matrix
                                results
        :param boundingRange: The allowable deviation from the values in
                              ``expectedResults``. Exceeding this range will fail
                              the test
    
        :returns: A sequence of results from the model; each result is a sequence of
          <Date string>, <Value>, <Likelihood Score>,  <Record Number>
        """
    
      lastRowId = len(labels)
      data = self.getModelResults(uid, lastRowId)
    
      # We get this back from the API backwards
      data.reverse()
    
      with open("results.%s.csv" % (metricName,), "w") as fh:
        writer = csv.writer(fh)
        writer.writerow(("timestamp", "metric_value", "likelihood", "rowid"))
        writer.writerows(data)
    
      # Retrieve results broadcast by Anomaly Service and compare against known
      # data
      anomalyServiceResults = self._reapAnomalyServiceResults(
        metricId=uid,
>       numRowsExpected=len(labels))

../htmengine/tests/integration/result_quality_test.py:384: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tests.integration.result_quality_test.ResultQualityTests testMethod=testRNSData>
metricId = 'ed0da9015f504a9fb6c8169a2e99294e', numRowsExpected = 4027

    def _reapAnomalyServiceResults(self, metricId, numRowsExpected):
      """ Retrieve likelihood results from our AMQP message queue that is bound to
        Anomaly Service's results fanout exchange
    
        NOTE that Anomaly Service fans out all results for all models via "fanout"
        exchange, so our queue might contain results from additional models, which
        we filter out.
    
        :param metricId: unique id of our metric/model
        :param numRowsExpected: number of result rows expected by caller
    
        :returns: a sequence of dicts conforming to the schema of the results items
          per model_inference_results_msg_schema.json
        """
      rows = []
    
>     @retry(timeoutSec=30, initialRetryDelaySec=0.5,
            retryExceptions=(AssertionError,))
      def getBatch(amqpClient):
        message = amqpClient.getOneMessage(self.resultsQueueName, noAck=False)
E       NameError: global name 'retry' is not defined

../htmengine/tests/integration/result_quality_test.py:294: NameError
```

**Comment by vitaly-krugl** [comment]
@SaganBolliger, wait ... it's not done yet -  the simple replacement of the internal retry implementation in test_case_base.py with the `retry` decorator from `nta.utils.error_handling` resulted in loss of functionality and changed the behavior, and has additional cleanup that needs to take place:

1. The original logged every iteration, which helped during debugging: `LOGGER.info("Attempt %i of %s: %s", i, f.__name__, firstLine)`
2. The original implementation paused `delay` seconds between retries, and that delay didn't grow. The way that the new parameters are configured will lead to some functions taking longer to complete than before, since the delay is allowed to double with each iteration)
3. Also

CC @scottpurdy, @oxtopus

**Comment by vitaly-krugl** [comment]
@SaganBolliger: When you submit code for review, the assumption is that it passed tests and pylint.

File: htm.it/htm/it/test_utils/app/test_case_base.py [hunk 1]
```diff
@@ -121,7 +91,8 @@ def fastCheckSequenceEqual(self, seq1, seq2):
         (i, seq1[i], seq2[i]))
 
 
-  @retry(duration=600, delay=5)
+  @retry(timeoutSec=600, initialRetryDelaySec=5, 
+        retryExceptions=(AssertionError,))
```
Comment (scottpurdy) [inline htm.it/htm/it/test_utils/app/test_case_base.py]: nit: indentation is off by a space on each of these

Comment (vitaly-krugl) [inline htm.it/htm/it/test_utils/app/test_case_base.py]: @scottpurdy, please see my feedback regarding `retryExceptions=(AssertionError,)` above.

**Comment by SaganBolliger** [comment]
@vitaly-krugl Sorry, I thought it did. Trying to figure out how this happened. I've reverted this PR.

**Comment by vitaly-krugl** [comment]
This happens to everyone every now and then. This is one of those changes that affects things in multiple sub-projects and takes some getting used to.

**Pull request merged**